RetroAchievements/RAWeb
# feat: sync GameAlternatives to game_sets, add double writes

This PR is prep work for making hubs a separate resource from games.

This is achieved by utilizing the `game_sets` table, which is meant to replace `GameAlternatives`.

As a result of this PR, `game_sets` and its related tables are now fully synchronized with `GameAlternatives`, with the intention that in future work we can drop `GameAlternatives` and switch over to `game_sets`.

A `GameSet` is a collection of games. Two `type`s of sets are supported: "hub" and "game". The "game" `type` allows `game_sets` to support the "Similar Games" relation.

**How to test this PR:**
```
sail artisan migrate
sail artisan ra:sync:game-sets
```
Observe in your database you have three tables that are fully populated: `game_sets`, `game_set_games`, and `game_set_links`. The links table is a new table introduced meant to facilitate hub<->hub relationships.

On the game page, when you add or remove related games/hubs from a game, you should see the values in these tables update. This can be confirmed by doing something like:
```php
$gameSet = GameSet::where("type", "game")
  ->where("game_id", 1)
  ->first();

$gamesCount = $gameSet->games->count();
$linksCount = $gameSet->linkedGameSets->count();

$sum = $gamesCount + $linksCount; // this value should change as you add and remove related games
```
